subscription_test.go

```diff
@@ -0,0 +1,196 @@
+package graphql_test
+
+import (
+	"context"
+	"encoding/json"
+	stdErrors "errors"
+	"testing"
+
+	graphql "github.com/graph-gophers/graphql-go"
+	"github.com/graph-gophers/graphql-go/errors"
+	"github.com/graph-gophers/graphql-go/gqltesting"
+)
+
+type rootResolver struct {
+	*helloResolver
+	*helloSaidResolver
+}
+
+type helloResolver struct{}
+
+func (r *helloResolver) Hello() string {
+	return "Hello world!"
+}
+
+var resolverErr = stdErrors.New("resolver error")
+
+type helloSaidResolver struct {
+	err      error
+	upstream <-chan *helloSaidEventResolver
+}
+
+type helloSaidEventResolver struct {
+	msg string
+	err error
+}
+
+func (r *helloSaidResolver) HelloSaid(ctx context.Context) (chan *helloSaidEventResolver, error) {
+	if r.err != nil {
+		return nil, r.err
+	}
+
+	c := make(chan *helloSaidEventResolver)
+	go func() {
+		for r := range r.upstream {
+			select {
+			case <-ctx.Done():
+				close(c)
+				return
+			case c <- r:
+			}
+		}
+		close(c)
+	}()
+
+	return c, nil
+}
+
+func (r *helloSaidEventResolver) Msg() (string, error) {
+	return r.msg, r.err
+}
+
+func closedUpstream(rr ...*helloSaidEventResolver) <-chan *helloSaidEventResolver {
```

Comment:
Should this be named `closedChannel` instead? For a moment, I got confused about the meaning of `upstream` here.